File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    buy := prices[0];
    profit := 0;

    for _, pr := range prices {
        buy = min(buy, pr)
        profit = max(profit, pr - buy)
    }

    return profit
}

func min(a int, b int) int {
    if a > b {
        return b
    }
    return a
}

func max(a int, b int) int {
    if a < b {
        return b
    }
    return a
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy = prices[0];
        int profit = 0;

        for (int pr : prices) {
            buy = min(buy, pr);
            profit = max(profit, pr - buy);
        }
        
        return profit;
    }
};